File: android/src/main/java/com/superwallreactnative/models/TriggerResult.kt
package com.superwallreactnative.models

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.ReadableMap
import com.superwall.sdk.models.triggers.TriggerResult

class TriggerResult {
  companion object {
    fun toJson(triggerResult: TriggerResult): ReadableMap {
      val map = Arguments.createMap()
      when (triggerResult) {
        is TriggerResult.PlacementNotFound -> {
          map.putString("result", "placementNotFound")
        }
        is TriggerResult.NoAudienceMatch -> {
          map.putString("result", "noAudienceMatch")
        }
        is TriggerResult.Paywall -> {
          map.putString("result", "paywall")
          map.putMap("experiment", Experiment.toJson(triggerResult.experiment))
        }
        is TriggerResult.Holdout -> {
          map.putString("result", "holdout")
          map.putMap("experiment", Experiment.toJson(triggerResult.experiment))
        }
        is TriggerResult.Error -> {
          map.putString("result", "error")
          map.putString("errorMessage", triggerResult.error.localizedMessage)
        }
      }
      return map
    }
  }
}

File: src/public/TriggerResult.ts
import { Experiment } from './Experiment';

export enum TriggerResultType {
  placementNotFound,
  noAudienceMatch,
  paywall,
  holdout,
  error,
}

// TypeScript class for TriggerResult
export class TriggerResult {
  type: TriggerResultType;
  experiment?: Experiment;
  error?: string;

  private constructor(
    type: TriggerResultType,
    experiment?: Experiment,
    error?: string
  ) {
    this.type = type;
    this.experiment = experiment;
    this.error = error;
  }

  static placementNotFound(): TriggerResult {
    return new TriggerResult(TriggerResultType.placementNotFound);
  }

  static noAudienceMatch(): TriggerResult {
    return new TriggerResult(TriggerResultType.noAudienceMatch);
  }

  static paywall(experiment: Experiment): TriggerResult {
    return new TriggerResult(TriggerResultType.paywall, experiment);
  }

  static holdout(experiment: Experiment): TriggerResult {
    return new TriggerResult(TriggerResultType.holdout, experiment);
  }

  static error(error: string): TriggerResult {
    return new TriggerResult(TriggerResultType.error, undefined, error);
  }

  static fromJson(json: any): TriggerResult {
    switch (json.result) {
      case 'placementNotFound':
        return TriggerResult.placementNotFound();
      case 'noAudienceMatch':
        return TriggerResult.noAudienceMatch();
      case 'paywall':
        return TriggerResult.paywall(Experiment.fromJson(json.experiment));
      case 'holdout':
        return TriggerResult.holdout(Experiment.fromJson(json.experiment));
      case 'error':
        return TriggerResult.error(json.error);
      default:
        throw new Error('Invalid TriggerResult type');
    }
  }
}
